### PR TITLE
Update API GUIDE to include multi host training and add some colors

### PR DESCRIPTION
**Pull request by JackCaoG**
My goal is to make `API_GUIDE` to be a good entry point for any firs time pytorch/xla user. We can put more technical details on different docs, but `API_GUIDE` should include a big picture.

rendered version in https://github.com/pytorch/xla/blob/JackCaoG/update_API_GUIDE/API_GUIDE.md#running-on-multiple-xla-devices-with-multi-processing.

### CODE REVIEW

File: API_GUIDE.md [hunk 1]
```diff
@@ -152,6 +160,25 @@ See the
...
+
+Now in order to run the same models on a TPU v4-16 (which has 2 host, each with 4 TPU devices), you will need to
+  - Make sure each host can access the training script and training data. This is usually done by using the `gcloud scp` command or `gcloud ssh` command to copy the training scripts to all hosts.
+  - Run the same training command on all hosts at the same time.
+
+```
+gcloud alpha compute tpus tpu-vm ssh $USER-pjrt --zone=us-central2-b --project=$PROJECT --worker=all --command="PJRT_DEVICE=TPU python3 train_mnist_xla.py"
+```
+
+Note that the model code and training scirpt is the same for the multi-process training and the multi-host training. PyTorch/XLA and the underlying infrastructure will make sure each device is aware of the global topology and each device's local and global ordinal. Cross-device communication will happen across all devices instead of local devices. For more details regarding PJRT runtime and how to run it on pod, please refer to this [doc](https://github.com/pytorch/xla/blob/master/docs/pjrt.md#tpu). For more information about PyTorch/XLA and TPU pod, please refer to this [guide](https://cloud.google.com/tpu/docs/pytorch-pods).
```
Comment (JackCaoG) [inline API_GUIDE.md:180]: @will-cromar I think we should update TPU pod guide to use PJRT than XRT? WDYT?

File: API_GUIDE.md [hunk 2]
```diff
@@ -135,11 +135,19 @@ if __name__ == '__main__':
...
 There are three differences between this multi-device snippet and the previous
-single device snippet:
-
-- `xmp.spawn()` creates the processes that each run an XLA device.
-- `MpDeviceLoader` loads the training data onto each device.
-- `xm.optimizer_step(optimizer)` consolidates the gradients between cores and issues the XLA device step computation.
+single device snippet. Let's go over then one by one.
+
+- `xmp.spawn()`
+  - Creates the processes that each run an XLA device
```
Comment (cowanmeg) [inline API_GUIDE.md:141]: XLA device -> XLA device.

File: API_GUIDE.md [hunk 3]
```diff
@@ -135,11 +135,19 @@ if __name__ == '__main__':
...
+  - Each process will only be able to access the device assigned to the current process. For example on a TPU v4-8, there will be 4 processes being spawn up and each process will own a TPU device.
+  - Note that if you print the `xm.xla_device()` on each process you will see `xla:0` on all devices. This is because each process can only see one device. This does not mean multi-process is not functioning.
+- `MpDeviceLoader`
+  - Loads the training data onto each device.
+  - `MpDeviceLoader` can wrap on a torch dataloader and overlap the dataloading with device execution to improve the performance.
+  - `MpDeviceLoader` also call `xm.mark_step` for you every `batches_per_execution`(default to 1) batch being yeild.
```
Comment (cowanmeg) [inline API_GUIDE.md:147]: yeild -> yield

File: API_GUIDE.md [hunk 4]
```diff
@@ -152,6 +160,25 @@ See the
 [full multiprocessing example](https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist.py)
 for more on training a network on multiple XLA devices with multi-processing.
 
+### Running on Multiple XLA Hosts
+Multi-host setup for different accelerators can be very different. This doc will talk about the device independent bits of multi-host training and will use the TPU + PJRT runtime(currently available on 1.13 and 2.x releases) as an example.
```
Comment (cowanmeg) [inline API_GUIDE.md:164]: Maybe add a one sentence description of what PJRT is and does.
Are there any changes in the earlier training code that are required to get it running on multiple hosts?

File: API_GUIDE.md [hunk 5]
```diff
@@ -152,6 +160,25 @@ See the
...
+
+Let's assume you have the above mnist example from above section in a `train_mnist_xla.py`. If it is a single host multi device training, you would run it like
+
+```
+PJRT_DEVICE=TPU python3 train_mnist_xla.py
+```
+
+Now in order to run the same models on a TPU v4-16 (which has 2 host, each with 4 TPU devices), you will need to
+  - Make sure each host can access the training script and training data. This is usually done by using the `gcloud scp` command or `gcloud ssh` command to copy the training scripts to all hosts.
```
Comment (cowanmeg) [inline API_GUIDE.md:173]: First mention of gcloud. Maybe add pointer about how to reserve TPUs here or at the beginning.

**Review by cowanmeg — COMMENTED** [review]
Left some small comments and edits.

File: API_GUIDE.md (hunk 4 above)
Comment (JackCaoG) [reply API_GUIDE.md:164]: I stated those(no code change, where to find PJRT doc) at the end of this section, do you think it is better to talk about hose in the beginning?

File: API_GUIDE.md (hunk 5 above)
Comment (JackCaoG) [reply API_GUIDE.md:173]: same, I share the link for TPU pod doc link at the end of this section, do you think it is better to move that up here?

File: API_GUIDE.md [hunk 6]
```diff
@@ -135,11 +135,19 @@ if __name__ == '__main__':
...
-single device snippet:
-
-- `xmp.spawn()` creates the processes that each run an XLA device.
-- `MpDeviceLoader` loads the training data onto each device.
-- `xm.optimizer_step(optimizer)` consolidates the gradients between cores and issues the XLA device step computation.
+single device snippet. Let's go over then one by one.
+
+- `xmp.spawn()`
+  - Creates the processes that each run an XLA device
+  - Each process will only be able to access the device assigned to the current process. For example on a TPU v4-8, there will be 4 processes being spawn up and each process will own a TPU device.
```
Comment (vanbasten23) [inline API_GUIDE.md:142]: By "TPU device", do you mean a TPU chip (containing 2 cores)?

Also, just checking if it's on a v3-8, would there be 8 processes instead?

Comment (JackCaoG) [reply API_GUIDE.md:142]: I felt like chip is too much of a detail. On V4-8 there are 4 devices, each of them is a chip. One v3-8 there are 8 devices, each of them is a core.

File: API_GUIDE.md [hunk 7]
```diff
@@ -152,6 +160,25 @@ See the
 [full multiprocessing example](https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist.py)
 for more on training a network on multiple XLA devices with multi-processing.
 
+### Running on Multiple XLA Hosts
+Multi-host setup for different accelerators can be very different. This doc will talk about the device independent bits of multi-host training and will use the TPU + PJRT runtime(currently available on 1.13 and 2.x releases) as an example.
+
+Let's assume you have the above mnist example from above section in a `train_mnist_xla.py`. If it is a single host multi device training, you would run it like
+
+```
```
Comment (vanbasten23) [inline API_GUIDE.md:168]: nit: maybe mention the user need to ssh into the TPU VM and run the command there?

Comment (JackCaoG) [reply API_GUIDE.md:168]: good point, I forgot to mentioned where to run these commands.

File: API_GUIDE.md (hunk 1 above)
Comment (will-cromar) [reply API_GUIDE.md:180]: Agreed

File: API_GUIDE.md (hunk 5 above)
Comment (alanwaketan) [reply API_GUIDE.md:173]: Yea, I think it's better to move the link about. Should we mention shared disk here?

**Review by alanwaketan — APPROVED** [review]
LGTM.

File: API_GUIDE.md [hunk 8]
```diff
@@ -135,11 +135,19 @@ if __name__ == '__main__':
...
-
-- `xmp.spawn()` creates the processes that each run an XLA device.
-- `MpDeviceLoader` loads the training data onto each device.
-- `xm.optimizer_step(optimizer)` consolidates the gradients between cores and issues the XLA device step computation.
+single device snippet. Let's go over then one by one.
+
+- `xmp.spawn()`
+  - Creates the processes that each run an XLA device
+  - Each process will only be able to access the device assigned to the current process. For example on a TPU v4-8, there will be 4 processes being spawn up and each process will own a TPU device.
+  - Note that if you print the `xm.xla_device()` on each process you will see `xla:0` on all devices. This is because each process can only see one device. This does not mean multi-process is not functioning.
```
[the remark this answers is not in the document]
Comment (JackCaoG) [reply API_GUIDE.md:143]: ah ok.. totally forgot about v3. Let me reword it.

File: API_GUIDE.md [hunk 9]
```diff
@@ -152,6 +160,25 @@ See the
 [full multiprocessing example](https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist.py)
 for more on training a network on multiple XLA devices with multi-processing.
 
+### Running on Multiple XLA Hosts
+Multi-host setup for different accelerators can be very different. This doc will talk about the device independent bits of multi-host training and will use the TPU + PJRT runtime(currently available on 1.13 and 2.x releases) as an example.
+
+Let's assume you have the above mnist example from above section in a `train_mnist_xla.py`. If it is a single host multi device training, you would run it like
```
[the remark this answers is not in the document]
Comment (JackCaoG) [reply API_GUIDE.md:166]: ok, I can add that

File: API_GUIDE.md (hunk 4 above)
Comment (cowanmeg) [reply API_GUIDE.md:164]: My bad, must have missed that. It's good where it is.

File: API_GUIDE.md (hunk 5 above)
Comment (JackCaoG) [reply API_GUIDE.md:173]: I will move the doc to the top, I think share disk is a bit too detial, we should probably mention that in the user guide on cloud dev site.

File: API_GUIDE.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (JackCaoG) [reply API_GUIDE.md:166]: I am having some trouble creating a pod to play with these command. For now I will leave this doc as "How it should work" and direct user to the pod user guide.. The major problem now is that user guide still uses XRT.

**Review by vanbasten23 — APPROVED** [review]
